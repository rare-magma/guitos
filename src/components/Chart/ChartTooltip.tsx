import Big from "big.js";
import { Col, Container, Row } from "react-bootstrap";
import { useConfig } from "../../context/ConfigContext";
import { intlFormat, roundBig } from "../../utils";

interface ChartTooltipProps {
  active?: boolean;
  payload?: {
    name: string;
    payload?: {
      id: string;
      item: string;
      name: string;
      type: string;
      value: number;
    };
    value: number;
    unit: string;
  }[];
  label?: string;
  key1?: string | undefined;
  key2?: string | undefined;
}

function ChartTooltip({
  active,
  payload,
  label,
  key1,
  key2,
}: ChartTooltipProps) {
  const { intlConfig } = useConfig();
  const currency = intlConfig?.currency;
  const showTooltip = active && payload?.length && currency;
  const filteredItemName = payload?.length && payload[0].payload?.item;
  const showMultipleLegends = showTooltip && payload.length > 1;

  const item1Value =
    showTooltip && intlFormat(roundBig(Big(payload[0].value), 2), currency);
  const item2Value =
    showMultipleLegends &&
    intlFormat(roundBig(Big(payload[1].value), 2), currency);

  return showTooltip ? (
    <Container className="m-2">
      <Row>{label}</Row>
      {showMultipleLegends ? (
        <>
          <Row className="me-1">
            <Col>{`${key1 ?? ""}:`}</Col>
            <Col className="text-end fixed-width-font">{item1Value}</Col>
          </Row>
          <Row className="me-1 flex-md-nowrap">
            <Col>{`${key2 ?? ""}:`}</Col>
            <Col className="text-end fixed-width-font col-md-auto">
              {item2Value}
            </Col>
          </Row>
        </>
      ) : (
        <Row className="me-1">
          <Col>{`${key1 ?? filteredItemName}:`}</Col>
          {key1 === "goal" ? (
            <Col className="text-end fixed-width-font">{`${payload[0].value}%`}</Col>
          ) : (
            <Col className="text-end fixed-width-font">{item1Value}</Col>
          )}
        </Row>
      )}
    </Container>
  ) : null;
}

export default ChartTooltip;
