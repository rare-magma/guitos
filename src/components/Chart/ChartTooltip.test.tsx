import { render, screen } from "@testing-library/react";
import { ChartTooltip } from "./ChartTooltip";

describe("ChartTooltip", () => {
  const comp = (
    <ChartTooltip
      active={true}
      label="label"
      payload={[{ name: "name", value: 123, unit: "$" }]}
    />
  );

  it("matches snapshot", () => {
    render(comp);
    expect(comp).toMatchSnapshot();
  });

  it("renders initial state", () => {
    render(comp);
    expect(screen.getByText("label")).toBeInTheDocument();
    expect(screen.getByText("$123.00")).toBeInTheDocument();
  });

  it("renders goal with %", () => {
    render(comp);
    render(
      <ChartTooltip
        active={true}
        label="goal"
        payload={[{ name: "goal", value: 123, unit: "%" }]}
        key1="goal"
      />,
    );
    expect(screen.getByText("123%")).toBeInTheDocument();
  });

  it("renders 2 legends", () => {
    render(comp);
    render(
      <ChartTooltip
        active={true}
        payload={[
          { name: "revenue", value: 456, unit: "$" },
          { name: "expenses", value: 789, unit: "$" },
        ]}
        key1="revenue"
        key2="expenses"
      />,
    );
    expect(screen.getByText("$456.00")).toBeInTheDocument();
    expect(screen.getByText("$789.00")).toBeInTheDocument();
  });
});
