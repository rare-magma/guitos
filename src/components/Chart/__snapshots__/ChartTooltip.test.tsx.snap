// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChartTooltip > matches snapshot 1`] = `
<ChartTooltip
  active={true}
  label="label"
  payload={
    [
      {
        "name": "name",
        "unit": "$",
        "value": 123,
      },
    ]
  }
/>
`;
